ANW-2218 support docker on production

thimios opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
<!--- Why is this change required? What problem does it solve? -->

## Related JIRA Ticket or GitHub Issue
<!--- Please link to the JIRA Ticket or GitHub Issue here: -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have authority to submit this code.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

mark-cooper commented on docker_prod/docker-compose.yml:
```diff
@@ -0,0 +1,52 @@
+      ASPACE_DB_MIGRATE: true
+      ASPACE_JAVA_XMX: "-Xmx2048m"
+      JAVA_OPTS: "-Djava.awt.headless=true -Dfile.encoding=UTF-8 -server -Xss1024k -Djavax.accessibility.assistive_technologies=''"
+      DB_ADDR: db
+      DB_NAME: archivesspace
+      DB_USER: as
+      DB_PASS: as123
+      SOLR_URL: http://solr:8983/solr/archivesspace
```
These are redundant / not actually used (The `APPCONFIG_` values are the real ones).

mark-cooper commented on docker_prod/docker-compose.yml:
```diff
@@ -0,0 +1,52 @@
+    cap_add:
+      - SYS_NICE
+    container_name: mysql
+    command: --character-set-server=utf8 --collation-server=utf8_unicode_ci --innodb_buffer_pool_size=1G --innodb_buffer_pool_instances=2 --log_bin_trust_function_creators=1
```
A reasonable default is probably `innodb_buffer_pool_instances=1` (that's a straightforward 1x1gb buffer pool vs. 2x512mb).

mark-cooper commented on docker_prod/docker-compose.yml:
```diff
@@ -0,0 +1,52 @@
+    image: archivesspace/solr:4.0.0-RC1
+    command: solr-create -p 8983 -c archivesspace -d archivesspace
+    environment:
+      SOLR_JAVA_MEM: "-Xms2g -Xmx2g"
```
Default `1g` is probably good. Many sites won't need more than that.

mark-cooper commented on docker_prod/docker-compose.yml:
```diff
@@ -0,0 +1,52 @@
+    ports:
+      - "3306:3306"
+    environment:
+      MYSQL_ROOT_PASSWORD: "123456"
+      MYSQL_DATABASE: archivesspace
+      MYSQL_USER: as
+      MYSQL_PASSWORD: as123
```
Recommend passwords should pass through from env. Something like:

```
MYSQL_ROOT_PASSWORD: ${ROOT_PASSWORD}
```

Or maybe require an env file?

```
env_file:
  - ./.env
```

mark-cooper commented on docker_prod/docker-compose.yml:
```diff
@@ -0,0 +1,52 @@
+      - db
+      - solr
+    environment:
+      APPCONFIG_DB_URL: "jdbc:mysql://db:3306/archivesspace?useUnicode=true&characterEncoding=UTF-8&user=as&password=as123&useSSL=false&allowPublicKeyRetrieval=true"
```
This should interpolate envvars (`password=$PASSWORD`) or require it be set via an env file? (more in db comments)

thimios commented on docker-compose-prod.yml:
```diff
@@ -0,0 +1,38 @@
+services:
+  app:
+    image: archivesspace/archivesspace:@TAG@
```
@mark-cooper The @TAG@ token is set by ant: https://github.com/archivesspace/archivesspace/pull/3420/files#diff-bffc59192cdccaf509753556461ff111ed11b8611616b7ebf7bbfd0700c3a711R318

to the git tag as passed to it from the github workflow:  https://github.com/archivesspace/archivesspace/pull/3420/files#diff-87db21a973eed4fef5f32b267aa60fcee5cbdf03c67fafdc2a9b553bb0b15f34R47

Which leaves this strange @TAG@ token in the compose file but serves our release workflow... I do not have a better idea for this...

mark-cooper replied:
I think it looks fine. A possible alternative might be to use `latest` as the default in place of `@TAG@` so it is runnable locally if you want to try it out easily (i.e. with `docker compose -f docker-compose.prod.yml up -d` for testing). But that exposes a potential issue in the case that "latest" isn't properly replaced during the build workflow (although then there could be checks for that etc.).

mark-cooper commented:
I haven't read up on everything so sorry if parts of the feedback are irrelevant (don't have time to deep dive r/n):

I think it broadly looks good. But there are a few things that are still needed. The major thing is data persistence. At least:

Additional host volume mounts for:

- the archivesspace data directory (something like: `./data:/archivesspace/data`)
- the mysql database (something like: `./database:/var/lib/mysql`)
- the solr data (something like: `./solr:/var/solr/data`)

I'm going off memory so paths may be somewhat off. I recommend host volume mounts for this use case because it's a bit more straightforwardly "visible" to anyone less familiar with Docker (vs. Docker managed volumes).

It may or may not be desirable to directly expose ports for the db and solr. A "safer" default would be to not expose the ports and just let ASpace connect to them.

IIRC the use case is simplify deployment as much as possible and in that case I think I'd recommend bundling an nginx proxy to listen on port 4000 (but configurable) and then also restrict the ASpace ports. The only exposed port then would be 4000 for whatever other proxy / load balancer is going to connect to ASpace, and it just needs to proxy pass / forward to 4000 as the paths would be predefined for the individual apps like: https://github.com/archivesspace/archivesspace/blob/master/proxy/default.conf.release

There should be instructions about backing stuff up that may need to be tailored to how the docker bundle is working. For example if the db ports are not exposed then ASpace can be used to run a SQL dump (assuming that script still works!). Restoring would be about destroying the container, deleting the `./database` files and putting the dump file in `./sql`.

mark-cooper commented:
@thimios on second thoughts I think host volume mounts are probably not a good idea for data. At the very least there were issues on Windows using them. Using Docker volumes should always work equivalently across operating systems. That would look something like:

```yml
services:
  app:
    # ...
    volumes:
      - appdata:/archivesspace/data
  db:
    # ...
    volumes:
      - dbdata:/var/lib/mysql
  solr:
    # ...
    volumes:
      - solrdata:/var/solr/data

volumes:
  appdata:
  dbdata:
  solrdata:
```

There would just need to be docs about how data will be lost if the volumes are removed and about the importance of backups (but that's as always).

If it helps you out I can make a PR to this branch with the update after testing across Windows, Mac and Ubuntu (I got them all). I can also add the proxy if you're interested in seeing that.

thimios commented:
@mark-cooper I actually started working on setting up external mounts but did not know about the windows issue... Sure it would be great if you would create a PR to this branch with the proposed changes and I can take care of the documentation

thimios commented:
@mark-cooper I added one more commit with the WIP changes I started doing after your feedback. I opted for a single .env file because I feel it is easier to manage and because I could not find a way to pass an environment variable to be used within the docker-compose file for the proxy port configuration without makind the standard command `docker compose up` more complex...

mark-cooper commented:
@thimios I pushed one commit directly to use the required mount path when using docker volumes. I've tested Ubuntu and Mac. I just need to test Windows tomorrow. I also wanted to look into an integrated backup container (if it seems doable quickly).

thimios commented:
@mark-cooper sounds great!

mark-cooper commented:
@thimios I've tested this on Windows now as well. It works as expected. I think it would be a good idea to review the Dockerfile before calling this actually ready for production in general but the concept is looking good (and that work can be a follow up as I think this PR is ready).

Docker desktop for Windows:

![Screenshot 2025-01-29 034309](https://github.com/user-attachments/assets/ecd0225e-98d5-4f74-a76a-228facc32d4a)

thimios commented:
@mark-cooper just squashed previous commits and added a new one, as I found a bug:

`⋊> ~/w/a/archivesspace on ANW-2218_docker_support_for_prod ◦ ./build/run build-docker-conf-package -Dversion="v4.0.0" -Ddocker-tag="4.0.0"                                                                                           
Buildfile: /home/thimios/work/archives-space/archivesspace/build/build.xml                                                                                                                                                                   
                                                                                                                                                                                                                                             
set-classpath:                                                                                                                                                                                                                               
                                                                                                                                                                                                                                             
build-docker-conf-package:                                                                                                                                                                                                                   
    [mkdir] Created dir: /home/thimios/work/archives-space/archivesspace/build/docker_prod/archivesspace/config                                                                                                                              
    [mkdir] Created dir: /home/thimios/work/archives-space/archivesspace/build/docker_prod/archivesspace/plugins                                                                                                                             
    [mkdir] Created dir: /home/thimios/work/archives-space/archivesspace/build/docker_prod/archivesspace/sql                                                                                                                                 
    [mkdir] Created dir: /home/thimios/work/archives-space/archivesspace/build/docker_prod/archivesspace/proxy-config                                                                                                                        
     [copy] Copying 1 file to /home/thimios/work/archives-space/archivesspace/build/docker_prod/archivesspace                                                                                                                                
     [copy] Copying 1 file to /home/thimios/work/archives-space/archivesspace/build/docker_prod/archivesspace                                                                                                                                
     [copy] Copying 1 file to /home/thimios/work/archives-space/archivesspace/build/docker_prod/archivesspace/proxy-config                                                                                                                   
      [zip] Building zip: /home/thimios/work/archives-space/archivesspace/archivesspace-docker-v4.0.0.zip                                                                                                                                    
                                                                                                                                                                                                                                             
BUILD SUCCESSFUL                                                                                                                                                                                                                             
Total time: 3 seconds`

because of replacing 'latest' in the docker compose file with the 'docker-tag' argument: https://github.com/archivesspace/archivesspace/pull/3420/files#diff-bffc59192cdccaf509753556461ff111ed11b8611616b7ebf7bbfd0700c3a711R323

we end up having also https://github.com/archivesspace/archivesspace/pull/3420/files#diff-0faee0a2e11c359bfbe8424e6a5c5a0d58881176f85af37e273497e0dad42840R32 replaced with '4.0.0' which doesn't work...

I switched back to using a token in the compose file and adding some documentation about it. Let me know what you think.

See also the updated docs here: https://github.com/archivesspace/tech-docs/pull/221

mark-cooper commented:
@thimios LGTM. The fact that it _can't_ work without token replacement is likely a feature, meaning you can't accidentally run a version that wasn't intended (easily).

The only objection I can think of is that it would be _nicer_ if the configuration could all be contained in `.env`. So the compose file would have:

```
image: archivesspace/archivesspace:${ARCHIVESSPACE_TAG}
```

With `ARCHIVESSPACE_TAG=@docker-tag@` in `.env` (`.env.docker.prod`). Obviously it would be even better if that can be set to the release version without requiring it to be set manually, but iirc you had a problem with doing that, so this works fine.

thimios commented:
@mark-cooper that is a good point, it is of course better to set the version through an env variable

mark-cooper commented:
@thimios LGTM.
